server/api/shifts/getAllShifts.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+$root = realpath($_SERVER["DOCUMENT_ROOT"]);
+require_once "$root/server/config.php";
+
+getAllShifts($_GET);
+
+/*
+ * The fields that CAN be returned are: shiftId, startTime, endTime, archived, createdAt, lastModifiedDate, siteId, createdBy, lastModifiedBy
+ *
+ * The data can be used to narrow down what is selected from the database:
+ * {
+ *   "shiftId": false,
+ *   "endTime": true,
+ *   "siteId": true
+ * }
+ *
+ * If the database field is missing it is assumed that you DON'T want it.
+ */
+function getAllShifts($data) {
+	GLOBAL $DB_CONN;
+	$defaultSelectColumns = array('shiftId', 'startTime', 'endTime', 'archived', 'createdAt', 'lastModifiedDate', 'siteId', 'createdBy', 'lastModifiedBy');
+	
+	// construct select columns list
+	$selectColumns = [];
+	if (!is_null($data) && !empty($data)) {
+		foreach ($defaultSelectColumns as $key) {
+			if (isset($data[$key])) {
+				if ($data[$key] == true) {
+					$selectColumns[] = $key; // append the select columns
+				}
+			}
+		}
+	}
+	$selectColumnsString = join(',', $selectColumns);
+
+	$year = 2018;
+	if(isset($data['year'])) {
+		$year = $data['year'];
+	}
+
+	$stmt = $DB_CONN->prepare('SELECT ' . $selectColumnsString . ' FROM Shift WHERE YEAR(startTime) = ?');
+	$stmt->execute(array($year));
+	$shifts = $stmt->fetchAll(PDO::FETCH_ASSOC);
+	foreach ($shifts as &$shift) {
```

Comment:
What's the point of the & in front of shift? Does this avoid you having to create a new array to pass back? Or what is happening here?

Reply:
> "In order to be able to directly modify array elements within the [foreach] loop precede $value with &. In that case the value will be assigned by reference." [PHP Manual](http://pl.php.net/manual/en/control-structures.foreach.php)

Reply:
Good stuff, I didn't know that. Thanks Collins